File: scrape_oppa.py
import requests
from bs4 import BeautifulSoup
import json
import re

# Oppa Design category URLs
urls = [
    "https://www.oppa.com.br/moveis/sofas/?PS=12",
    "https://www.oppa.com.br/moveis/poltronas/?PS=12"
]

products = []

headers = {
    'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36'
}

for url in urls:
    print(f"Scraping category: {url}")
    try:
        response = requests.get(url, headers=headers)
        soup = BeautifulSoup(response.text, 'html.parser')
        
        # This will depend on VTEX or Magento structure, let's try generic links first or find VTEX pattern
        # Many Brazilian e-commerces use VTEX.
        items = soup.find_all('a', class_=re.compile(r'product-image|shelf-item__image|product__image', re.I))
        if not items:
            items = soup.select('.prateleira a') # VTEX
            
        for item in items[:4]: # Take top 4 from each category to get 8 total
            href = item.get('href')
            if not href.startswith('http'):
                href = 'https://www.oppa.com.br' + href
                
            print(f"Found product: {href}")
            
            # Now scrape the product page
            p_res = requests.get(href, headers=headers)
            p_soup = BeautifulSoup(p_res.text, 'html.parser')
            
            name = p_soup.find('h1')
            name = name.text.strip() if name else "Unknown"
            
            price_elem = p_soup.find(class_=re.compile(r'bestPrice|skuBestPrice|preco-por|price__best', re.I))
            price = price_elem.text.strip() if price_elem else "R$ 0,00"
            
            img_elem = p_soup.find('img', id='image-main')
            if not img_elem:
                img_elem = p_soup.find('img', class_=re.compile(r'sku-rich-image-main|product-image__main', re.I))
            img_url = img_elem.get('src') if img_elem else ""
            
            # Extract dimensions from description or specs table
            # VTEX specs table
            specs = p_soup.find_all('td', class_='value-field')
            spec_labels = p_soup.find_all('th', class_='name-field')
            
            width, height, depth = None, None, None
            
            for th, td in zip(spec_labels, specs):
                label = th.text.strip().lower()
                val = td.text.strip().lower()
                if 'largura' in label:
                    m = re.search(r'([\d.,]+)', val)
                    if m: width = float(m.group(1).replace(',', '.'))
                elif 'altura' in label:
                    m = re.search(r'([\d.,]+)', val)
                    if m: height = float(m.group(1).replace(',', '.'))
                elif 'profundidade' in label:
                    m = re.search(r'([\d.,]+)', val)
                    if m: depth = float(m.group(1).replace(',', '.'))

            products.append({
                "name": name,
                "price": price,
                "url": href,
                "image": img_url,
                "category": "sofa" if "sofa" in url else "poltrona",
                "dimensions": {
                    "width_cm": width,
                    "height_cm": height,
                    "depth_cm": depth
                }
            })
            
    except Exception as e:
        print(f"Error scraping {url}: {e}")

with open('oppa_products.json', 'w', encoding='utf-8') as f:
    json.dump(products, f, indent=2, ensure_ascii=False)
    
print(f"Saved {len(products)} products to oppa_products.json")

File: scrape_oppa.mjs
import puppeteer from 'puppeteer';
import fs from 'fs/promises';

async function scrapeOppa() {
    const browser = await puppeteer.launch({ headless: 'new' });
    const page = await browser.newPage();

    await page.setViewport({ width: 1280, height: 800 });
    await page.setUserAgent('Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36');

    const urls = [
        "https://www.oppa.com.br/collections/sofa",
        "https://www.oppa.com.br/collections/poltrona"
    ];

    const allProducts = [];

    for (const url of urls) {
        console.log(`Loading category: ${url}`);
        await page.goto(url, { waitUntil: 'networkidle2', timeout: 60000 });

        // Scroll to load lazy items
        await page.evaluate(() => window.scrollBy(0, 1000));
        await new Promise(r => setTimeout(r, 2000));

        // Get product links (Shopify uses /products/)
        const productLinks = await page.evaluate(() => {
            const links = Array.from(document.querySelectorAll('a'));
            return links
                .map(a => a.href)
                .filter(href => href.includes('/products/') && !href.includes('/collections/'))
                .filter((href, index, self) => self.indexOf(href) === index) // Unique
                .slice(0, 4); // Take 4 per category
        });

        console.log(`Found ${productLinks.length} products on ${url}`);

        for (const link of productLinks) {
            console.log(`Scraping product: ${link}`);
            try {
                await page.goto(link, { waitUntil: 'networkidle2', timeout: 60000 });

                const product = await page.evaluate(() => {
                    const nameEl = document.querySelector('h1.product-single__title') || document.querySelector('h1');
                    const name = nameEl ? nameEl.innerText.trim() : "Unknown";

                    // Try to find price
                    const priceEls = document.querySelectorAll('.product__price, .price-item, [class*="price"]');
                    let price = "Price not found";
                    for (let el of priceEls) {
                        if (el.innerText && el.innerText.includes('R$')) {
                            price = el.innerText.trim();
                            const match = price.match(/R\$\s*[\d\.,]+/);
                            if (match) price = match[0];
                            break;
                        }
                    }

                    // Try to find main image
                    const imgEl = document.querySelector('.product-single__photo img') ||
                        document.querySelector('img.lazyautosizes') ||
                        document.querySelector('.product-image-main img') ||
                        document.querySelector('img[src*="cdn.shopify.com"]');
                    // Get highest res version by replacing size modifiers like _400x400
                    let image = imgEl ? imgEl.src : "";
                    if (image.startsWith('//')) image = 'https:' + image;
                    if (image) image = image.replace(/_\d+x\d+/, '');

                    let width = null, height = null, depth = null;

                    // Look through page text for dimensions
                    // Oppa puts them in a structured table or text list in the description
                    const rawText = document.body.innerText.toLowerCase();
                    const wMatch = rawText.match(/largura[^\d]*([\d.,]+)/);
                    if (wMatch) width = parseFloat(wMatch[1].replace(',', '.'));

                    const hMatch = rawText.match(/altura[^\d]*([\d.,]+)/);
                    if (hMatch) height = parseFloat(hMatch[1].replace(',', '.'));

                    const dMatch = rawText.match(/profundidade[^\d]*([\d.,]+)/);
                    if (dMatch) depth = parseFloat(dMatch[1].replace(',', '.'));

                    return { name, price, image, width, height, depth };
                });

                allProducts.push({
                    name: product.name,
                    price: product.price,
                    url: link,
                    image: product.image,
                    category: url.includes('sofa') ? "sofa" : "poltrona",
                    dimensions: {
                        width_cm: product.width,
                        height_cm: product.height,
                        depth_cm: product.depth
                    }
                });
            } catch (err) {
                console.log(`Failed to scrape ${link}: ${err.message}`);
            }
        }
    }

    await fs.writeFile('oppa_products_with_dims.json', JSON.stringify(allProducts, null, 2));
    console.log(`✅ Saved ${allProducts.length} products.`);
    await browser.close();
}

scrapeOppa().catch(console.error);
